### PR TITLE
[serve] properly broadcast autoscaling config when updated

### PR DESCRIPTION
**Pull request by zcin**
[serve] properly broadcast autoscaling config when updated

When `metrics_interval_s` or `look_back_period_s` is modified in the autoscaling config, it should be properly broadcasted to the actor that is collecting autoscaling metrics.

Signed-off-by: Cindy Zhang <cindyzyx9@gmail.com>

---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/ray-project/ray/pull/42661).
* #42578
* __->__ #42661

### CODE REVIEW

File: python/ray/serve/_private/replica.py [hunk 1]
```diff
@@ -165,6 +165,21 @@ def __init__(
...
+
+    def start(self):
+        """Start periodic background tasks."""
+        self._metrics_pusher.start()
+
+    def shutdown(self):
+        """Stop periodic background tasks."""
+        self._metrics_pusher.shutdown()
+
+    def set_autoscaling_config(self, autoscaling_config: AutoscalingConfig):
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/replica.py:178]: Can the `autoscaling_config` argument be `None`? The `if self._autoscaling_config` conditional below implies that's the case.

Could you change the `autoscaling_config` type hint to `Optional[AutoscalingConfig]` if so?

File: python/ray/serve/_private/test_utils.py [hunk 2]
```diff
@@ -95,6 +97,46 @@ def check_telemetry_not_recorded(storage_handle, key):
     )
 
 
+def get_num_running_replicas(name: str, app_name: str = SERVE_DEFAULT_APP_NAME) -> int:
+    """Get the replicas currently running for given deployment"""
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/test_utils.py:101]:
```suggestion
    """Get the replicas currently running for the given deployment."""
```

File: python/ray/serve/_private/test_utils.py [hunk 3]
```diff
@@ -95,6 +97,46 @@ def check_telemetry_not_recorded(storage_handle, key):
     )
 
 
+def get_num_running_replicas(name: str, app_name: str = SERVE_DEFAULT_APP_NAME) -> int:
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/test_utils.py:100]:
```suggestion
def get_num_running_replicas(deployment_name: str, app_name: str = SERVE_DEFAULT_APP_NAME) -> int:
```

File: python/ray/serve/_private/version.py [hunk 4]
```diff
@@ -175,7 +176,11 @@ def _get_serialized_options(
                 reconfigure_dict[option_name] = getattr(
                     self.deployment_config, option_name
                 )
-                if isinstance(reconfigure_dict[option_name], BaseModel):
+                if isinstance(reconfigure_dict[option_name], AutoscalingConfig):
+                    reconfigure_dict[option_name] = reconfigure_dict[option_name].dict(
+                        include={"metrics_interval_s", "look_back_period_s"}
+                    )
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/version.py:182]: Why do we have this logic? Are people expected to pass in an `AutoscalingConfig` through their `reconfigure` method?

Comment (zcin) [reply python/ray/serve/_private/version.py:182]: The logic here is generic version handling. It is not specific to user's `reconfigure` method. Depending on what changed in the updated deployment config, the deployment state manager may push the new config to replicas, or simply update some state within deployment state manager. If `metrics_itnerval_s` or `look_back_period_s` is changed in the autoscaling config, we want to push those changes to the replica.

Comment (shrekris-anyscale) [reply python/ray/serve/_private/version.py:182]: Oh I see, thanks for explaining! I'm not too familiar with `include=` in Pydantic. Does that line mean that _only_ if `metrics_interval_s` or `look_back_period_s` changes, we push those changes to the replica? Or would we broadcast the changes if other fields in the `AutoscalingConfig` change too?

**Review by edoakes — APPROVED** [review]

File: python/ray/serve/_private/version.py (hunk 4 above)
Comment (zcin) [reply python/ray/serve/_private/version.py:182]: Yup, that's exactly right, we only broadcast if those two fields change.

**Review by shrekris-anyscale — APPROVED** [review]

**Pull request merged**